src/exe/Main.hs

```diff
@@ -224,18 +254,47 @@ application = do
 processInputFile
   :: Members '[Report, Embed IO] r => Options -> FilePath -> Sem r ()
 processInputFile opts inputFile = do
-  input <- embed $ readFile inputFile
+  input    <- embed $ readFile inputFile
+  frontend <- parseFrontend (optFrontend opts)
+  let (output, modName) = processInput input opts frontend
+--      inputModule = HSE.fromParseResult (HSE.parseModule input
+--      intermediateModule =
+--        evalPM (processModule (FromHSE.transformModule inputModule)) state
+--      outputModule = ToHSE.transformModule inputModule intermediateModule
+  case optOutputDir opts of
+    Just outputDir -> do
+      let outputFile = outputDir </> makeOutputFileName inputFile modName
+      embed $ createDirectoryIfMissing True (takeDirectory outputFile)
+      embed $ writeFile outputFile output -- (prettyPrintModule outputModule)
+    Nothing -> embed $ putStrLn output --(prettyPrintModule outputModule)
+
+-- | Processes a given module as a string with the corresponding front end.
+--   Returns the output as a string and the module name.
+processInput :: String -> Options -> Frontend -> (String, Maybe String)
+processInput input opts frontend = case frontend of
+  HSE    -> processInputHSE input opts
+  GHClib -> error "Not yet implemented"
+
+-- | Processes a given module as a string with the @haskell-src-extensions@
+--   front end.
+processInputHSE :: String -> Options -> (String, Maybe String)
+processInputHSE input opts =
   let state       = initPMState opts
       inputModule = HSE.fromParseResult (HSE.parseModule input)
       intermediateModule =
-        evalPM (processModule (FromHSE.transformModule inputModule)) state
+          evalPM (processModule (FromHSE.transformModule inputModule)) state
       outputModule = ToHSE.transformModule inputModule intermediateModule
-  case optOutputDir opts of
-    Just outputDir -> do
-      let outputFile = outputDir </> makeOutputFileName inputFile inputModule
-      embed $ createDirectoryIfMissing True (takeDirectory outputFile)
-      embed $ writeFile outputFile (prettyPrintModule outputModule)
-    Nothing -> embed $ putStrLn (prettyPrintModule outputModule)
+  in  (prettyPrintModuleHSE outputModule, moduleNameHSE inputModule)
```

Comment:
The only difference between the `HSE` and `GHClib` front end will probably be parsing and pretty printing. Thus, the call to `processModule` could be moved to `processInputFile`.

Reply:
I'm resolving this conversation since the suggested change is blocked by #41 as you've correctly noted in today's meeting.